### PR TITLE
nixos/moinmoin: init module

### PR DESCRIPTION
**Pull request by mmilata**
##### Motivation for this change

Migrating my wiki instance from CentOS->NixOS.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @abbradar @danieldk

### CODE REVIEW

File: nixos/tests/all-tests.nix
Comment (danieldk) [inline nixos/tests/all-tests.nix]: Typo?

File: nixos/modules/services/web-apps/moin.nix
Comment (danieldk) [inline nixos/modules/services/web-apps/moin.nix]: This overrides `moin` from the `moinmoin` package. I am not sure, but I think in general it is good to avoid such global installs. But you might want to wait for @abbradar 's comments before changing this. (I do see the benefits of a 'moin' command with the right user and configuration set.)

File: nixos/modules/services/web-apps/moin.nix
Comment (danieldk) [inline nixos/modules/services/web-apps/moin.nix]: *I think* that if you remove this and add `plugin = "python2"` to the moin vassal configuration, the same uwsgi instance/emperor can be used for Python 2 and Python 3. (untested)

**Review by danieldk — COMMENTED** [review]
Thanks for implementing this! I do not use MoinMoin in farm mode (I have a separate vassal for each site), so I can only provide limited feedback.

File: nixos/modules/services/web-apps/moin.nix
Comment (danieldk) [inline nixos/modules/services/web-apps/moin.nix]: Oh, by the way, any reason for calling this module `moin` rather than `moinmoin`?

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/moin.nix]: @mmilata or you could leave it hardcoded the way you have it now, taking advantage of `systemd` features to manage your state directory and @danieldk could make a symlink. Dealers choice :smile:

Note that currently you are using `tmpfiles.d` to provision directories that `StateDirectory` could be used to provision if you leave the data directory hardcoded to `/var/lib/*`.

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/moin.nix]: If this package provides commands required to manage the application then generally that is seen as justification to add the package to the system, especially in the case where specific configuration is required.

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: Right, this is mostly leftover from me trying to figure out how to pass configuration directory to the wiki engine.

However the way it's implemented has the benefit that uwsgi is restarted whenever wiki configuration is changed. Is there a way to do this with uwsgi directly so there would be no reason to generate the file?

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: From the feedback on my previous module PR I figured it's preferred to have `dataDir` hardcoded under `/var/lib`. You can also bind-mount the external storage as `/var/lib/moin`, this way you should be able to configure systemd to wait for the mount before starting the wiki.

@aanderse I'm a fan of the systemd sandboxing options that `StateDirectory` enables, however this module doesn't have it's own unit file as everything is run under uWSGI daemon. Maybe using e.g. `spawn-fcgi` instead of uWSGI would be better in this regard, the wiki supports FastCGI as well, not sure about performance though.

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: Removing this results in the module pulling in uwsgi package that is built without any plugins. I'll wrap the option in `mkDefault` so that adding more plugins is easier.

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: I honestly don't know which name is preferred. The project seems to use "MoinMoin" the most, however not lowercase "moinmoin". They also use "moin" for source archive, github repo, CLI etc so I went with that. Happy to change it if matching the package name is more important.

**Comment by mmilata** [comment]
Thanks for the quick feedback guys!

Regarding the farmconfig, I currently only run single wiki instance so I don't need it, though easy way of running multiple instances seemed like a nice thing to have. I see several possibilities here:
* switching to vassal-per-instance if @danieldk wants to use or even help maintain this module
* switching to private uWSGI instance, or some other application server, to take advantage of systemd sandboxing
* forcing single instance, as that's what most users want and it's still quite easy to run multiple instances using NixOS containers

What do you think?

Changes:
* fix `all-tests.nix`
* wrapped `services.uwsgi.plugins` in `mkDefault`, removed setting `services.uwsgi.moin.instance.vassals.moin.pythonpath` as it had no effect

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/moin.nix]: @mmilata ah sorry. I just gave the module a quick look over and missed that fact. :+1:

**Comment by mmilata** [comment]
I've updated the module to use _gunicorn_ instead of _uWSGI_ as an application server. Personally, gunicorn seems easier to use than uWSGI, and it is possible to use systemd sandboxing options. Haven't enabled `DynamicUser` because I don't know how to make it work with the command line `moin` utility.

Changes:
* farmconfig setup is no longer used, multiple wikis are implemented through multiple gunicorn systemd units
* default `forceSSL=true` and `enableACME=true` for nginx vhosts
* generated wrapper script no longer shadows the shipped `moin` script

@danieldk @aanderse PTAL

**Comment by mmahut** [comment]
@GrahamcOfBorg test moin

File: nixos/modules/services/web-apps/moin.nix
Comment (mmahut) [inline nixos/modules/services/web-apps/moin.nix]: I think this can be removed as it's the default option.

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/moin.nix]: From an extremely quick glance at the site it isn't made overly clear which. For consistency sake, though, maybe we should go with `moinmoin`?

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/moin.nix]: Can you explain specifically what prompted you to add this, given the `tmpfiles` declarations below?

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/moin.nix]: Is there any concept of 'migrations' in `moinmoin` which might have some problems introduced during version upgrades here (particularly the `C` entries in your `tmpfiles`)? While I do like the declarative nature of `tmpfiles` I have concerns that `tmpfiles` may not be flexible enough **if** upgrading between versions of `moinmoin` requires special handling.

I'm not familiar with `moinmoin` as a sysadmin, so if my question is of no concern please disregard.

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/moin.nix]: If you didn't use `su` but instead just ran the command then nixos sysadmins could have better control over the access to this script with `sudo` rules. The sysadmin can always type `sudo -u moin-ExampleWiki` to be explicit.

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/moin.nix]: I assume that `moinmoin` doesn't actually go ahead and create these users... it might be worth a quick mention that the sysadmin must create the account first.

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/moin.nix]: Extremely minor nitpick... spacing.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: I added this in order for the service to have read-write access to the directory when using `ProtectSystem="strict"`. IMO it's more self-descriptive than using `ReadWritePaths=`.

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: The `data` dir is supposed to be migrated to newer version using
`moin --config-dir=/path/to/config_dir --wiki-url=wiki.example.org/ migration data`.
Not sure whether we should do this automatically, however the `C` should be fine here for manual migrations.

Regarding the `underlay` dir the docs suggest to replace it with the one from the newer version. We can make `preStart` always replace it at the cost of slower startup. Symlinking doesn't work unfortunately.

References:
* https://github.com/moinwiki/moin-1.9/blob/master/docs/README.migration
* https://master19.moinmo.in/HelpOnUpdating

File: nixos/modules/services/web-apps/moin.nix
Comment (mmilata) [reply nixos/modules/services/web-apps/moin.nix]: I wanted to make setting up simple instance as easy as possible, not sure whether a script that just adds single argument to the `moin` cli helps with that. For maximum flexibility the `moin` cli is added to `systemPackages` as well.

Or did you mean to use sudo with some default rules instead?

**Review by mmilata — COMMENTED** [review]
Changes:
* removed redundant default attributes
* `flip map` -> `forEach`
* renamed module to `moinmoin`
* moved underlay initialization from tmpfiles to preStart
* added note about creating user accounts

PTAL @aanderse

**Comment by mmahut** [comment]
@GrahamcOfBorg test moin

**Comment by mmilata** [comment]
@GrahamcOfBorg test moinmoin

Renamed the test as well:)

**Comment by mmahut** [comment]
@GrahamcOfBorg test moinmoin

**Comment by mmahut** [comment]
This looks good to me, @aanderse a final word?

**Review by aanderse — APPROVED** [review]

File: nixos/modules/services/web-apps/moin.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/moin.nix]: Personally I don't like the approach taken here (a script using `su`) and my preference would certainly be to document setting up access via `sudo` rules. I want to be clear on this point what I state is only my **opinion**, and not a hard requirement preventing this PR from being merged. You may take either approach.

**Comment by aanderse** [comment]
> This looks good to me, @aanderse a final word?

Looks like a few things need to be fixed up maybe (bot is failing) but I have no major show stoppers on the quality of the code, only opinions that @mmilata is free to follow or disregard. Overall looks good :+1:

**Comment by mmahut** [comment]
```tests.moin``` failed as I run the wrong test name, kicking the test on aarch64 where we run out of space.

@GrahamcOfBorg test moinmoin

**Comment by mmilata** [comment]
Thx for the review everyone!

**Pull request merged**
